File: app/src/main/java/com/kira/kdownloader/settings/platform/LanguageManager.java
package com.kira.kdownloader.settings.platform;

import android.app.LocaleManager;
import android.content.Context;
import android.content.ContextWrapper;
import android.content.res.Configuration;
import android.os.Build;
import android.os.LocaleList;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Locale;
import java.util.Objects;

public final class LanguageManager {
    public static final List<Language> SUPPORTED=Collections.unmodifiableList(Arrays.asList(
            new Language("","Follow system"),new Language("en","English"),new Language("es","Español"),
            new Language("fr","Français"),new Language("de","Deutsch"),new Language("pt","Português"),
            new Language("ru","Русский"),new Language("ar","العربية"),new Language("hi","हिन्दी"),
            new Language("id","Bahasa Indonesia"),new Language("ja","日本語"),new Language("zh","中文")));
    private LanguageManager(){}
    public static final class Language{private final String tag,display;public Language(String tag,String display){this.tag=tag;this.display=display;}public String getTag(){return tag;}public String getDisplay(){return display;}@Override public boolean equals(Object o){if(this==o)return true;if(!(o instanceof Language))return false;Language x=(Language)o;return tag.equals(x.tag)&&display.equals(x.display);}@Override public int hashCode(){return Objects.hash(tag,display);}@Override public String toString(){return "Language(tag="+tag+", display="+display+")";}}
    public static String displayName(String tag){for(Language language:SUPPORTED)if(language.tag.equals(tag))return language.display;if(!tag.trim().isEmpty())return Locale.forLanguageTag(tag).getDisplayName();return "Follow system";}
    public static void apply(Context context,String tag){if(Build.VERSION.SDK_INT>=Build.VERSION_CODES.TIRAMISU){LocaleManager manager=context.getSystemService(LocaleManager.class);if(manager!=null)manager.setApplicationLocales(tag.trim().isEmpty()?LocaleList.getEmptyLocaleList():LocaleList.forLanguageTags(tag));}}
    public static Context wrap(Context base,String tag){if(Build.VERSION.SDK_INT>=Build.VERSION_CODES.TIRAMISU||tag.trim().isEmpty())return base;Locale locale=Locale.forLanguageTag(tag);Locale.setDefault(locale);Configuration configuration=new Configuration(base.getResources().getConfiguration());configuration.setLocale(locale);return new ContextWrapper(base.createConfigurationContext(configuration));}
}

File: app/src/main/java/com/kira/kdownloader/settings/platform/LanguageManager.kt
package com.kira.kdownloader.settings.platform

import android.app.LocaleManager
import android.content.Context
import android.content.ContextWrapper
import android.content.res.Configuration
import android.os.Build
import android.os.LocaleList
import java.util.Locale

/**
 * Applies the app's language preference (Section 8).
 *
 * On Android 13+ the per-app language is delegated to the system [LocaleManager] so it is honoured
 * consistently and shows up in system settings. On older versions the stored tag is applied by
 * wrapping the base context's configuration in `attachBaseContext`.
 */
object LanguageManager {

    /** An empty tag means "Follow system". */
    data class Language(val tag: String, val display: String)

    val SUPPORTED = listOf(
        Language("", "Follow system"),
        Language("en", "English"),
        Language("es", "Español"),
        Language("fr", "Français"),
        Language("de", "Deutsch"),
        Language("pt", "Português"),
        Language("ru", "Русский"),
        Language("ar", "العربية"),
        Language("hi", "हिन्दी"),
        Language("id", "Bahasa Indonesia"),
        Language("ja", "日本語"),
        Language("zh", "中文"),
    )

    fun displayName(tag: String): String =
        SUPPORTED.firstOrNull { it.tag == tag }?.display
            ?: tag.takeIf { it.isNotBlank() }?.let { Locale.forLanguageTag(it).displayName }
            ?: "Follow system"

    /** Applies the language at runtime. On <33 the caller must also recreate the activity. */
    fun apply(context: Context, tag: String) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            val lm = context.getSystemService(LocaleManager::class.java)
            lm?.applicationLocales = if (tag.isBlank()) {
                LocaleList.getEmptyLocaleList()
            } else {
                LocaleList.forLanguageTags(tag)
            }
        }
        // On <33 nothing to do here; [wrap] applies the locale to newly created contexts.
    }

    /** Wraps [base] with the configured locale. Used from `attachBaseContext` on API < 33. */
    fun wrap(base: Context, tag: String): Context {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU || tag.isBlank()) return base
        val locale = Locale.forLanguageTag(tag)
        Locale.setDefault(locale)
        val config = Configuration(base.resources.configuration)
        config.setLocale(locale)
        return ContextWrapper(base.createConfigurationContext(config))
    }
}
